iothub/device/src/Messaging/IncomingMessage.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// The data structure that represents the message that is received from IotHub.
+    /// </summary>
+    public class IncomingMessage
+    {
+        private readonly byte[] _payload;
+
+        internal IncomingMessage(byte[] payload)
+        {
+            _payload = payload;
+        }
+
+        /// <summary>
+        /// An identifier for the message used for request-reply patterns.
+        /// Format: A case-sensitive string ( up to 128 char long) of ASCII 7-bit alphanumeric chars
+        /// + {'-', ':', '/', '\', '.', '+', '%', '_', '#', '*', '?', '!', '(', ')', ',', '=', '@', ';', '$', '''}.
+        /// Non-alphanumeric characters are from URN RFC.
+        /// </summary>
+        public string MessageId
```

Comment:
fields have public getters and internal setters

Reply:
Should all be `protected internal` for unit testing and mocking purposes. And copy the <remarks> that Nick/Nate put on their models.

Reply:
Updated internal fields to protected internal.